### PR TITLE
[AAP-38778] chore: improve jwt expired exception logging

### PR DESCRIPTION
**Pull request by Dostonbek1**
# [AAP-38778]  Improve JWT expired exception logging

## Description

- What is being changed?
Add more meaningful log message when jwt token is expired.
- Why is this change needed?
To make it easier for customers/developers to understand the root cause of authentication errors.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [x] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [ ] I have added relevant comments to complex code sections
- [ ] I have updated documentation where needed
- [ ] I have considered the security impact of these changes
- [ ] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

### Steps to Test
1. 
2. 
3. 

### Required Actions

- [ ] Requires documentation updates
  
- [ ] Requires downstream repository changes
  
- [ ] Requires infrastructure/deployment changes
  
- [ ] Requires coordination with other teams
  
- [ ] Blocked by PR/MR: #XXX
  

### Screenshots/Logs

image

### CODE REVIEW

File: ansible_base/jwt_consumer/common/auth.py [hunk 1]
```diff
@@ -170,18 +171,23 @@ def validate_token(self, unencrypted_token, decryption_key):
...
-            self.log_and_raise(_("JWT has expired"))
+            expired_token = self.decode_jwt_token(
+                unencrypted_token,
+                decryption_key,
+                options={"require": local_required_field, "verify_exp": False},
+            )
+            expired_time = expired_token.get("exp")
+            now = datetime.now().timestamp()
+            time_diff = int(now - expired_time)
+            self.log_and_raise(_(f"JWT has expired {time_diff} seconds ago - check for clock skew"))
```
Comment (kdelee) [inline ansible_base/jwt_consumer/common/auth.py]: fwiw from my perspective, its worth extra cost of doing this decoding a second time because if we are in this situation -- this is quite bad. 

Either
* clocks are out of sync
* component is so backlogged that tokens are arriving at it already expired
* some other bad/unsupported case of re-using old jwt tokens

**Comment by john-westcott-iv** [comment]
@Dostonbek1 please let us know if you need reviews once the tests are all fixed.

**Review by kdelee — APPROVED** [review]
Having this error in way that is exposed to end user is big improvement/should reduce time to diagnoses of the root cause. We still need to do some work in aapqa-provisioner to provision nodes w/ clocks in sync and MAYBE consider if installer should check this.

File: ansible_base/jwt_consumer/common/auth.py [hunk 2]
```diff
@@ -173,26 +180,31 @@ def map_user_fields(self):
...
 
         # Decrypt the token
         try:
             logger.info("Decrypting token")
-            validated_body = jwt.decode(
+            validated_body = self.decode_jwt_token(
                 unencrypted_token,
                 decryption_key,
-                audience="ansible-services",
                 options={"require": local_required_field},
```
Comment (john-westcott-iv) [inline ansible_base/jwt_consumer/common/auth.py]: Should we move this to be the default in `decode_jwt_token`? That method could take an argument like "additional options" and do an `options.update(additional_options)`.

File: ansible_base/jwt_consumer/common/auth.py [hunk 3]
```diff
@@ -143,9 +146,13 @@ def parse_jwt_token(self, request):
 
         logger.info(f"User {self.user.username} authenticated from JWT auth")
 
-    def log_and_raise(self, conditional_translate_object, expand_values={}):
+    def log_and_raise(self, conditional_translate_object, expand_values={}, error_code=None):
         logger.error(conditional_translate_object.not_translated() % expand_values)
-        raise AuthenticationFailed(conditional_translate_object.translated() % expand_values)
+        translated_error_message = conditional_translate_object.translated() % expand_values
+        if error_code == 498:
+            raise InvalidTokenException(translated_error_message)
```
Comment (john-westcott-iv) [inline ansible_base/jwt_consumer/common/auth.py:153]: Just to confirm, you've tested and this exception type gets properly picked up in the login flow? From a quick glance it does not look like anything is requiring AuthenticationFaield exception above this call.

File: ansible_base/jwt_consumer/common/auth.py [hunk 4]
```diff
@@ -143,9 +146,13 @@ def parse_jwt_token(self, request):
 
         logger.info(f"User {self.user.username} authenticated from JWT auth")
 
-    def log_and_raise(self, conditional_translate_object, expand_values={}):
+    def log_and_raise(self, conditional_translate_object, expand_values={}, error_code=None):
         logger.error(conditional_translate_object.not_translated() % expand_values)
-        raise AuthenticationFailed(conditional_translate_object.translated() % expand_values)
+        translated_error_message = conditional_translate_object.translated() % expand_values
+        if error_code == 498:
```
Comment (john-westcott-iv) [inline ansible_base/jwt_consumer/common/auth.py]: 498 is used in a couple places, do you want to make a constant incase we ever want to change this number?

File: ansible_base/jwt_consumer/common/auth.py (hunk 2 above)
Comment (Dostonbek1) [reply ansible_base/jwt_consumer/common/auth.py]: Refactored.

File: ansible_base/jwt_consumer/common/auth.py (hunk 3 above)
Comment (Dostonbek1) [reply ansible_base/jwt_consumer/common/auth.py:153]: Here are results from a live server testing with these changes:
```
Feb 25 15:56:15 ip-10-0-0-54 automation-eda-controller[92302]: 2025-02-25 15:56:15,591 ansible_base.jwt_consumer.common.auth ERROR    JWT expired 788 seconds ago - check for clock skew. Request ID: a7b69775-f179-41ad-8cd6-0ede0d2bafe5
```
Screenshot 2025-02-25 at 10 58 26 AM

Screenshot 2025-02-25 at 10 58 39 AM

**Review by tznamena — APPROVED** [review]

**Review by john-westcott-iv — APPROVED** [review]
